thoth-core/src/main/java/fr/maif/eventsourcing/impl/KafkaEventPublisher.java

```diff
@@ -60,30 +67,49 @@ public KafkaEventPublisher(ActorSystem system, ProducerSettings<String, EventEnv
         Source<EventEnvelope<E, Meta, Context>, SourceQueueWithComplete<EventEnvelope<E, Meta, Context>>> queue = Source.queue(queueBufferSize1, OverflowStrategy.backpressure());
 
         RunnableGraph<Pair<SourceQueueWithComplete<EventEnvelope<E, Meta, Context>>, Source<EventEnvelope<E, Meta, Context>, NotUsed>>> tmpVar = queue
-                .toMat(BroadcastHub.of((Class<EventEnvelope<E, Meta, Context>>)e.getClass(), 256), Keep.both());
+                .toMat(BroadcastHub.of((Class<EventEnvelope<E, Meta, Context>>) e.getClass(), 256), Keep.both());
 
         Pair<SourceQueueWithComplete<EventEnvelope<E, Meta, Context>>, Source<EventEnvelope<E, Meta, Context>, NotUsed>> pair = tmpVar
                 .run(materializer);
-
         this.kafkaProducer = producerSettings.createKafkaProducer();
         this.producerSettings = producerSettings.withProducer(this.kafkaProducer);
         this.queue = pair.first();
         this.eventsSource = pair.second();
     }
 
-    public void start(EventStore<?, E, Meta, Context> eventStore) {
+    public <TxCtx> void start(EventStore<TxCtx, E, Meta, Context> eventStore, ConcurrentReplayStrategy concurrentReplayStrategy) {
+
         RestartSource
                 .onFailuresWithBackoff(
                         restartInterval,
                         maxRestartInterval,
                         0,
                         () -> {
                             LOGGER.info("Starting/Restarting publishing event to kafka on topic {}", topic);
-                            return eventStore.loadEventsUnpublished()
-                                    .via(publishToKafka(eventStore, Flow.<ProducerMessage.Results<String, EventEnvelope<E, Meta, Context>, EventEnvelope<E, Meta, Context>>>create().grouped(1000).map(List::ofAll)))
+                            return Source.completionStage(eventStore.openTransaction().toCompletableFuture())
+                                    .flatMapConcat(tx -> {
+
+                                        LOGGER.info("Replaying not published in DB for {}", topic);
+                                        ConcurrentReplayStrategy strategy = Objects.isNull(concurrentReplayStrategy) ? NO_STRATEGY : concurrentReplayStrategy;
```

Comment:
I would go for WAIT as default since it's the safer regarding event ordering, however everything except NO_STRATEGY is ok for me